temporalio/ext/src/client.rs

```diff
@@ -237,18 +246,59 @@ impl Client {
         self.invoke_rpc(service, callback, call)
     }
 
-    pub fn update_metadata(&self, headers: HashMap<String, String>) -> Result<(), Error> {
+    pub fn update_metadata(&self, headers: RHash) -> Result<(), Error> {
+        let ruby = Ruby::get().expect("Ruby not available");
+        let headers = partition_grpc_headers(&ruby, headers)?;
+        self.core
+            .get_client()
+            .set_headers(headers.headers)
+            .map_err(|err| error!("Invalid headers: {}", err))?;
         self.core
             .get_client()
-            .set_headers(headers)
-            .map_err(|err| error!("Invalid headers: {}", err))
+            .set_binary_headers(headers.binary_headers)
+            .map_err(|err| error!("Invalid headers: {}", err))?;
+        Ok(())
     }
 
     pub fn update_api_key(&self, api_key: Option<String>) {
         self.core.get_client().set_api_key(api_key);
     }
 }
 
+pub(crate) struct GrpcHeaders {
+    headers: HashMap<String, String>,
+    binary_headers: HashMap<String, Vec<u8>>,
+}
+
+fn partition_grpc_headers(ruby: &Ruby, hash: RHash) -> Result<GrpcHeaders, Error> {
+    let mut headers = HashMap::new();
+    let mut binary_headers = HashMap::new();
+    hash.foreach(|key: String, value: RString| {
+        if key.ends_with("-bin") {
+            if value.enc_get() != ruby.ascii8bit_encindex() {
+                return Err(Error::new(
+                    ruby.exception_arg_error(),
+                    format!("Value for metadata key {key} must be ASCII-8BIT"),
+                ));
+            }
+            binary_headers.insert(key, unsafe { value.as_slice() }.to_vec());
```

Comment:
It might make sense to move `to_vec` into unsafe block to indicate prevent ever accidentally holding onto the slice reference.
```suggestion
            binary_headers.insert(key, unsafe { value.as_slice().to_vec() });
```

Reply:
Makes sense. I tend to take the "smallest amount of unsafe possible", but it makes sense here. If I need another commit, I'll make this change too.